src/dune_rules/workspace.ml

```diff
@@ -479,7 +487,10 @@ module Context = struct
              | None ->
                User_error.raise
                  ~loc:loc_switch
-                 [ Pp.textf "Generated context name %S is invalid" name
+                 ?hints:(name_hint_opt name)
```

Comment:
There's an opportunity to refactor to use `Context_name.of_string_user_error` instead of `of_string_opt` (see `Stringlike_intf`).
Then you can use the `Stringlike` API to implement that hint, by defining a `hint_valid` function (which can use your basename implementation if the string looks like an absolute path, or something less elegant otherwise)

Reply:
After investigating, I will finally keep the error like this. Indeed, if I take this path, the error would look like: 
```
File "dune-workspace", line 5, characters 12-45:
5 |     (switch /path/to/the/opam/switch/named/dune)))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: "/path/to/the/opam/switch/named/dune" is an invalid context name.
It must not be empty, be the exact string "log", contain slashes, contain
backslashes or start with the symbol  ".".
Hint: (name dune) would be a correct context name
```
It makes the error message more obscure as it totally eclipse the context where it happens...

Reply:
Sorry, what's specifically wrong about that error message? You don't have to define `description_of_valid_string` if it doesn't make sense.

Reply:
Yes, I can remove it, but, IMHO, this is not the problem here. If we remove the `match ... with` we lost the information about the fact that the user should add `(name ...)` field in case we don't have it in the `(opam (switch ...))` context.

I don't know a way to specify this using the `Stringlike` API as the `hint_valid` name, insert a suggestion into an already existing string: `Format.sprintf "%s would be a correct context name" (f_extracted_from_hint_valid s)`.  I attempt to modify the `hint_valid` function with my `name_hint_opt` function, but the result does not look informative enough (as you can see in the output above).

Furthermore, I don't see a way to specify the additional message, as it was suggested in the original discussion on Slack:
```
The generated context name "..." from the switch name is invalid, Please specify a context name manually with the (name ..) field.
```
This message makes sens because the user has the information that the name what generated from the `switch` and that it should manually add the `(name ...)` field. Maybe I'm misunderstanding how the `Stringlike` API work and there is a more elegant way to do it? :confused:

Reply:
let's scratch that idea, it's just an implementation detail